### PR TITLE
EZP-30823: [7.x] Isolated cache during transactions to avoid race condition 

### PR DESCRIPTION
**Pull request by andrerom**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30823](https://jira.ez.no/browse/EZP-30823)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | `7.5`+, but depends on #2740 being merged first as this is for conflict resolution during merge
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

During attempt to backport the transaction isolation fixes done in #2703 a few cases where we should go a bit further appeared:
- Avoid stale data within transactions:
    - Mark items that has deferred delete/invalidation as a "cache miss" when retrieved from shared pool
    - Merge in in-memory handing in order to make sure in-memory cache is always cleared directly and not deferred to avoid using stal
- Avoid creating potentially stale data in shared cache pool:
    - Handle save during transaction => map to a deferred delete by key
    - Cache should only be updated when outer transaction finishes, inner transactions don't show up for other processes.
    - Rollbacks should not commit changes. And in SQL they break out of any level transaction by default, align with that
- BENEFIT: Take advantage of this to get rid of full cache clearing during rollback

REVIEW NOTES:
- The two Decorators we had has been merged into 1 in order to be able to handle in-memory cache during transaction changes

REVIEW QUESTION:
- Should we keep `ezpublish.cache_pool.inner` and instead remove the decoration?
- Ideally should have something to test this across processes, maybe something like a test with transaction which calls on another process to check it's value while still within transaction. If so, how and where?

### CODE REVIEW

File: eZ/Publish/Core/Persistence/Cache/Adapter/TransactionalInMemoryCacheAdapter.php [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+namespace eZ\Publish\Core\Persistence\Cache\Adapter;
+
+use Symfony\Component\Cache\Adapter\TagAwareAdapterInterface;
+use Psr\Cache\CacheItemInterface;
+use Symfony\Component\Cache\CacheItem;
+
+/**
+ * Internal proxy adapter invalidating our isolated in-memory cache, and defer shared pool changes during transactions.
+ *
+ * @intenral For type hinting inside eZ\Publish\Core\Persistence\Cache\*. For external, type hint on TagAwareAdapterInterface.
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Cache/Adapter/TransactionalInMemoryCacheAdapter.php]: typo:
```suggestion
 * @internal For type hinting inside eZ\Publish\Core\Persistence\Cache\*. For external, type hint on TagAwareAdapterInterface.
```

File: eZ/Publish/Core/Persistence/Cache/Adapter/TransactionalInMemoryCacheAdapter.php [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        $this->sharedPool = $sharedPool;
+        $this->inMemoryPools = $inMemoryPools;
+        $this->transactionDepth = $transactionDepth;
+        $this->deferredTagsInvalidation = empty($deferredTagsInvalidation) ? [] : \array_fill_keys($deferredTagsInvalidation, true);
+        $this->deferredItemsDeletion = empty($deferredItemsDeletion) ? [] : \array_fill_keys($deferredItemsDeletion, true);
+        // To modify protected $isHit when items are a "miss" based on deferred delete/invalidation during transactions
+        $this->setCacheItemAsMiss = \Closure::bind(
+            static function (CacheItem $item) {
+                // ... Might not work for anything but new items
+                $item->isHit = false;
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Cache/Adapter/TransactionalInMemoryCacheAdapter.php]: Uh, this one gave me a headache. Too bad Symfony does this exactly like that. Not to mention that PhpStorm shows this as erroneous...
No action required, just complaining about quality of Sf...

File: eZ/Publish/Core/Persistence/Cache/Adapter/TransactionalInMemoryCacheAdapter.php [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+        array $deferredTagsInvalidation = [],
+        array $deferredItemsDeletion = []
+    ) {
+        $this->sharedPool = $sharedPool;
+        $this->inMemoryPools = $inMemoryPools;
+        $this->transactionDepth = $transactionDepth;
+        $this->deferredTagsInvalidation = empty($deferredTagsInvalidation) ? [] : \array_fill_keys($deferredTagsInvalidation, true);
+        $this->deferredItemsDeletion = empty($deferredItemsDeletion) ? [] : \array_fill_keys($deferredItemsDeletion, true);
+        // To modify protected $isHit when items are a "miss" based on deferred delete/invalidation during transactions
+        $this->setCacheItemAsMiss = \Closure::bind(
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Cache/Adapter/TransactionalInMemoryCacheAdapter.php]: Seems this class member is missing.

File: eZ/Publish/Core/Persistence/Cache/Adapter/TransactionalInMemoryCacheAdapter.php [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+use Symfony\Component\Cache\Adapter\TagAwareAdapterInterface;
+use Psr\Cache\CacheItemInterface;
+use Symfony\Component\Cache\CacheItem;
+
+/**
+ * Internal proxy adapter invalidating our isolated in-memory cache, and defer shared pool changes during transactions.
+ *
+ * @intenral For type hinting inside eZ\Publish\Core\Persistence\Cache\*. For external, type hint on TagAwareAdapterInterface.
+ *
+ * @todo Naming? Is this something like IsolatedCacheDecorator
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Cache/Adapter/TransactionalInMemoryCacheAdapter.php]: TBH `IsolatedCacheAdapter` or `IsolatedCacheAdapterDecorator`, if you must. But I have no problem with `TransactionalInMemoryCacheAdapter`.

File: eZ/Publish/Core/Persistence/Cache/Adapter/TransactionalInMemoryCacheAdapter.php [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+            if ($this->itemIsDeferredMiss($item)) {
+                $fnSetCacheItemAsMiss($item);
+            }
+        }
+
+        return $iteratedItems;
+    }
+
+    /**
+     * @param CacheItem $item
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Cache/Adapter/TransactionalInMemoryCacheAdapter.php]: Nitpick: FQCN.
```suggestion
     * @param \Symfony\Component\Cache\CacheItem $item
```

File: eZ/Bundle/EzPublishCoreBundle/Resources/config/cache.yml [hunk 6]
```diff
@@ -2,7 +2,7 @@ parameters:
     ezpublish.cache_pool.factory.class: eZ\Bundle\EzPublishCoreBundle\ApiLoader\CacheFactory
 
 services:
-    ezpublish.cache_pool.inner:
+    ezpublish.cache_pool:
```
[the remark this answers is not in the document]
Comment (andrerom) [reply eZ/Bundle/EzPublishCoreBundle/Resources/config/cache.yml]: To be clear: you suggest to keep it like proposed here right? Or?

[the remark this answers is not in the document]
Comment (alongosz) [reply eZ/Bundle/EzPublishCoreBundle/Resources/config/cache.yml]: > To be clear: you suggest to keep it like proposed here right? Or?

Yep, let's keep it as it is now :+1: I need to find time in the future to investigate better decorating.

**Comment by andrerom** [comment]
@webhdx What's your POV/REview here?

**Pull request merged**